x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DeprecationRoleDescriptorConsumer.java

```diff
@@ -0,0 +1,203 @@
+        this(clusterService, threadPool, new DeprecationLogger(logger));
+    }
+
+    // package-private for testing
+    DeprecationRoleDescriptorConsumer(ClusterService clusterService, ThreadPool threadPool, DeprecationLogger deprecationLogger) {
+        this.deprecationLogger = deprecationLogger;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        // this String Set keeps "<date>-<role>" pairs so that we only log a role once a day.
+        this.dailyRoleCache = Collections.newSetFromMap(new LinkedHashMap<String, Boolean>() {
+            @Override
+            protected boolean removeEldestEntry(final Map.Entry<String, Boolean> eldest) {
+                return false == eldest.getKey().startsWith(todayISODate());
+            }
+        });
+        this.workQueue = new LinkedList<>();
+        this.workerBusy = false;
+    }
+
+    @Override
+    public synchronized void accept(Collection<RoleDescriptor> effectiveRoleDescriptors) {
+        final Collection<RoleDescriptor> uncachedEffectiveRoleDescriptors = new ArrayList<>();
+        for (RoleDescriptor roleDescriptor : effectiveRoleDescriptors) {
+            if (dailyRoleCache.add(buildCacheKey(roleDescriptor))) {
+                uncachedEffectiveRoleDescriptors.add(roleDescriptor);
+            }
+        }
+        if (false == uncachedEffectiveRoleDescriptors.isEmpty()) {
+            workQueue.add(uncachedEffectiveRoleDescriptors);
+            if (false == workerBusy) {
+                workerBusy = true;
+                try {
+                    // spawn another worker on the generic thread pool
+                    threadPool.generic().execute(() -> {
+                        while (true) {
+                            // synchronize access for the next work item
+                            final Collection<RoleDescriptor> workItem;
+                            synchronized (this) {
+                                workItem = workQueue.poll();
+                                if (workItem == null) {
+                                    workerBusy = false;
+                                    break;
+                                }
+                            }
+                            // executing the check asynchronously will not conserve the generated deprecation response headers (which is
+                            // what we want, because it's not the request that uses deprecated features, but rather the role definition.
+                            // Furthermore, due to caching, we can't reliably associate response headers to every request).
+                            logDeprecatedPermission(workItem);
+                        }
+                    });
+                } catch (RejectedExecutionException e) {
+                    workerBusy = false;
+                    logger.warn("Failed enqueue work for role deprecation messages");
+                    throw e;
```

Comment:
do we really need to rethrow the exception here? I mean should we fail a request because a deprecation log message was rejected by the threadpool?

Reply:
no, sure not! this was an oversight!